vulture_os/toolkit/api_parser/ubika/ubika.py

```diff
@@ -0,0 +1,279 @@
+                msg = f"[{__parser__}][__execute_query]: Connection failed (HTTPError)"
+                logger.exception(f'{msg}', extra={'frontend': str(self.frontend)})
+                raise e
+
+            if resp.status_code not in [200, 201]:
+                msg = f"[{__parser__}][__execute_query]: Status code not in [200, 201]. Code {resp.status_code} -- {resp.content}"
+                logger.exception(f'{msg}', extra={'frontend': str(self.frontend)})
+                if resp.status_code == 401:
+                    self.login()
+                retry -= 1
+
+            # grpc_error_code = resp_json.get('code')
+            # if grpc_error_code:
+                # grpc_error_msg = resp_json.get('message')
+                # if grpc_error_code == 16 and grpc_error_msg == "token has invalid claims: token is expired":
+                    # self.login()
+
+            break # no error we break from the loop
+
+        return resp_json
+
+
+    def get_logs(self, url: str, since: datetime, to: datetime) -> list:
+        logs = []
+        next_token = ""
+        last_save_token = ""
+
+        data = {
+            "pagination.pageSize": 10001,
+            "pagination.realtime": True
+        }
+
+        page_end = False
+        while not page_end:
+            if next_token:
+                data["pagination.pageToken"] = next_token
+            else:
+                data["filters.fromDate"] = int(since.timestamp() * 1000),
+                data["filters.toDate"] = int(to.timestamp() * 1000)
+
+            resp = self.__execute_query(
+                method = "GET",
+                url = url,
+                data = data)
+
+            next_token = resp.get('spec', {}).get('nextPageToken', '')
+            logger.debug(f"[{__parser__}][get_logs]: Paginate with url = {url}, last_save_token = {last_save_token}, next_token = {next_token}", extra={'frontend': str(self.frontend)})
+
+            if next_token != last_save_token:
+                newLogs = resp.get('spec', {}).get('items', [])
+                if newLogs:
+                    logs.extend(newLogs)
+                last_save_token = next_token
+            else:
+                page_end = True
+
+        return logs
+
+    def format_log(self, log: dict) -> str:
+        request_headers : list = log.get('request', {}).get('headers', [])
+        new_request_headers = {}
+        if request_headers:
+            for header in request_headers:
+                new_key = header.get('key').lower()
+                new_value = header.get('value')
+                if new_key and new_value:
+                    new_request_headers[new_key] = new_value
+        log['request']['headers'] = new_request_headers
+
+        request_cookies : list = log.get('request', {}).get('cookies', [])
+        new_request_cookies = {}
+        if request_cookies:
+            for cookie in request_cookies:
+                new_key = cookie.get('key').lower()
```

Comment:
Same problem: using `lower()` requires `cookie.get('key')` to be a string

```suggestion
                new_key = cookie.get('key', '').lower()
```